README
NAME
    Makefile::Parser - A simple parser for Makefiles

VERSION
    This document describes Makefile::Parser 0.215 released on 18 August
    2011.

SYNOPSIS
      use Makefile::Parser;

      $parser = Makefile::Parser->new;

      # equivalent to ->parse('Makefile');
      $parser->parse or
          die Makefile::Parser->error;

      # get last value assigned to the specified variable 'CC':
      print $parser->var('CC');

      # get all the variable names defined in the Makefile:
      @vars = $parser->vars;
      print join(' ', sort @vars);

      @roots = $parser->roots; # Get all the "root targets"
      print $roots[0]->name;

      @tars = $parser->targets;  # Get all the targets
      $tar = join("\n", $tars[0]->commands);

      # get the default target, say, the first target
      # defined in Makefile:
      $tar = $parser->target;

      $tar = $parser->target('install');
      # get the name of the target, say, 'install' here:
      print $tar->name;

      # get the dependencies for the target 'install':
      @depends = $tar->depends;

      # access the shell command used to build the current target.
      @cmds = $tar->commands;

      # parse another file using the same Parser object:
      $parser->parse('Makefile.old') or
        die Makefile::Parser->error;

      # get the target who is specified by variable EXE_FILE
      $tar = $parser->target($parser->var('EXE_FILE'));

DESCRIPTION
    This is a simple parser for Makefiles. At this very early stage, the
    parser only supports a limited set of features, so it may not recognize
    most of the advanced features provided by certain make tools like GNU
    make. Its initial purpose is to provide basic support for another module
    named Makefile::GraphViz, which is aimed to render the building process
    specified by a Makefile using the amazing GraphViz library. The Make
    module is not satisfactory for this purpose, so I decided to build one
    of my own.

    WARNING!!! This stuff is highly experimental and is currently at
    pre-alpha stage, so production use is strongly discouraged. Right now
    it's just a toy for parsing trivial makefiles.

    IMPORTANT!!! If you're looking for something more serious for parsing
    GNU makefiles, please see Makefile::Parser::GmakeDB instead. The GmakeDB
    parser has passed 51% of GNU make's official test suite as of this
    writing.

    If you're looking for something that can parse makefiles losslessly,
    take a look at the Makefile::DOM module which may fit your needs.

  SYNTAX SUPPORTED
    The current parser implementation has been trying to support a common
    feature set of both MS NMAKE and GNU make. In the future, different
    formats of Makefiles will be handled by individual subclasses such as
    Makefile::Parser::Gmake.

    Variable Definition
            MIN_T_FILES = $(PAT_COVER_FILES) t\optest.t t\my_perl.exe.t t\types.cod.t \
                t\catln.t t\exe2hex.t t\hex2bin.t t\bin2hex.t t\bin2asm.t t\ndisasmi.t \
                t\Idu.t t\pat_tree.t t\state_mac.t t\Idu-Util.t t\cidu.t \
                t\opname.t t\error.t t\operand.t t\01disasm.t t\02disasm.t t\03disasm.t \
                t\disasm_cover.t t\ndisasm.t
            T_FILES = t\main.cod.t t\bin2hex.exe.t t\hex2bin.exe.t $(MIN_T_FILES)
            DIRFILESEP = ^\

        "Simply expanded" variables' definition sytax in GUN make is also
        supported:

            FOO := blah blah blah

        which is considered invalid in Win32 NMake. "Recursively expanded"
        variables are currently treated as "simply expanded" variables.

        Variable redefinition can be handled as well:

            CC = cl

            %.obj : %.c
                $(CC) /nologo /c $<

            CC = gcc

            %.o : %.c
                $(CC) -c $<

        Variable expansion sytax

            ${abc}

        is accepted, whereas Win32 NMAKE will complain about it.

        Currently, environment variables defined in the command-line are not
        imported.

        I have no idea what default value should be assigned to built-in
        variables like $(MAKE) and $(CC). Currently they will be left
        untouched if they're not set explicitly in the Makefile.

        Due to the current implementation, expansion of unrecognized
        built-in varaibles and variables not previously defined by Makefile
        will NOT be performed. This behavior is different from any practial
        make tools, but is reasonable at this early stage of this parser.

    Explicit Rules
            $(CIDU_DLL) : C\idu.obj C\idu.def
                link /dll /nologo /debug /out:$@ /def:C\idu.def C\idu.obj

            $(CIDU_LIB) : $(CIDU_DLL)

            C\idu.obj : C\idu.c C\idu.h
                cd C
                cl /nologo /c /I . idu.c
                cd ..

            smoke : all pat_cover t\pat_cover.t \
                    t/pat_cover.ast.ast
                perl util\run-smoke.pl . smoke.html
                perl txt2html.pl t\*.t t\*.ast

            clean:
                copy t\pat_cover.ast.ast.html ..\ /Y
                $(RM_F) encoding.html encoding.pod state_mac.xml encoding.ast \
                    pat_tree.ast state_mac.ast \
                    main.cod pat_cover.pod pat_cover.html types.cod \
                    hex2bin.exe hex2bin.obj

        Specital variable $@ will be expanded using its value in the
        context.

    Implicit Rules

        Pattern Rules
                %.obj : %.asm
                    masm /t $<;

                %.exe : %.obj
                    link /BATCH /NOLOGO $<;

            The special varaibles $< and $* will be expanded according to
            the context.

        Old-Fashioned Suffix Rules
            Currently only double-suffix rules are supported:

                .SUFFIXES: .obj .asm .exe

                .asm.obj :
                    masm /t $<

                .obj.exe :
                    link /nologo $<

            At this moment, .SUFFIXES is a no-op. So any suffix-like things
            will be treated as suffixes, excluding the following example:

                .c.o: foo.h
                        $(CC) -c $(CFLAGS) $(CPPFLAGS) -o $@ $<

            In suffix rules, no prerequisites are allowed according to most
            make tools.

    Substitution References
            objects = foo.o bar.o baz.o
            sources = $(objects:.o=.c)  # foo.c bar.c baz.c

    Functions
        Currently the following GNU make makefile builtin functions are
        supported:

        " $(subst from,to,text) "
        " $(patsubst pattern,replacement,text) "
        " $(strip string) "
        " $(findstring find,text) "
        " $(filter pattern...,text) "
        " $(filter-out pattern...,text) "
        " $(sort list) "
        " $(word n,text) "
        " $(words text) "
        " $(wordlist s,e,text) "
        " $(firstword names...) "
        " $(lastword names...) "
        " $(dir names...) "
        " $(notdir names...) "
        " $(suffix names...) "
        " $(basename names...) "
        " $(addsuffix suffix,names...) "
        " $(addprefix prefix,names...) "
        " $(join list1,list2) "
        " $(wildcard pattern...) "
        " $(realpath names...) "
        " $(abspath names...) "
        " $(if condition,then-part[,else-part]) "
        " $(or condition1[,condition2[,condition3...]]) "
        " $(and condition1[,condition2[,condition3...]]) "
        " $(foreach var,list,text) "
        " $(error argument...) "
        " $(warning argument...) "
        " $(info argument...) "
        " $(shell cmd...) "

        Please consult the GNU make Manual for details and also take a look
        at the following file for some use cases:

        <http://github.com/agentzh/makefile-dom-pm/tree/master/t/gmake/sanit
        y/func-refs.t>

    Commands after ';'
            all : ; echo 'hello, world!'

        Specital variable $@ will be expanded using its value in the
        context.

    For the list of features which will be added very soon, take a look at
    the "TODO" section.

The Makefile::Parser Class
    This class provides the main interface to the Makefile parser.

  METHODS
    "$obj = Makefile::Parser->new()"
        It's the constructor for the Parser class. You may provide the path
        of your Makefile as the argument which . It is worth mentioning that
        the constructor will *not* call ->parse method internally, so please
        remember calling ->parse after you construct the parser object.

    "$obj->parse()"
    "$obj->parse($Makefile_name)"
    "$obj->parse($Makefile_name, { var => value, ... })"
        This method parse the specified Makefile (default to 'Makefile').

        When an error occurs during the parsing procedure, "parse" will
        return undef. Otherwise, a reference to Parser object itself is
        returned. It is recommended to check the return value every time you
        call this method. The detailed error info can be obtained by calling
        the "error" method.

        You can also pass a hash reference to specify initial variables and
        their values. Note that these variables are treated as "defaults" so
        assignments in the makefile have higher priority.

    "$obj->error()"
        It returns the error info set by the most recent failing operation,
        such as a parsing failure.

    "$obj->var($variable_name)"
        The var method returns the value of the given variable. Since the
        value of variables can be reset multiple times in the Makefile, so
        what you get is always the last value set to the variable. It's
        worth noting that variable reassignment can be handled appropriately
        during parsing since the whole parsing process is a one-pass
        operation compared to the multiple-pass strategy used by the CPAN
        module Make.

    "@vars = $obj->vars"
        This will return all the variables defined in the Makefile. The
        order may be quite different from the order they appear in the
        Makefile.

    "$obj->target($target_name)"
        This method returns a Makefile::Target object with the name
        specified. It will returns undef if the rules for the given target
        is not described in the Makefile. It is worth noting that only
        targets with a definition body will be considered as a *target*
        here.

        When $target_name is omitted, this method will return the default
        target, say, the first target defined in Makefile, to the user. This
        can be handy if you try to build a make tool on top of this module.

        It is important not to send something like "$(MY_LIB)" as the target
        name. Only raw values are acceptable. If you really want to do
        something like this, please use the following code:

            my $tar = $parser->target($parser->var('MY_LIB'));


Changes
Revision history for Perl extension Makefile::Parser.

0.216 2014-11-18
* fixed test failures in pod.t. thanks Olof Johansson for the patch.
* fixed test failures in simplemake.t with Gnu make 3.82.
* fixed warnings while running the build/test system. thanks Alberto
  Simões for the report.

0.215 2011-08-18
* fixed RT #41595: makesimple error: Can't locate object method
  "last_element" via package "MDOM::Token::Whitespace". thanks
  raleighr3 and jean.

0.214 2011-08-18
* fixed a bug when dealing with trailing backslashes in variable
  values. this fixed RT #69319: Error parsing line continuations.
  thanks Marcelo.

0.213 2011-08-17
* applied the patch from Philip Allison and fixed a bug
  regarding clearing initial set makefile variables before
  parsing makefiles.

0.212 2011-08-17
* Updated the author name, author email address, and code repository in the documentation.
* Updated the bundled Module::Install to the latest version, 1.01.

0.211 2008-03-16
* Added pointer to Makefile::DOM to Makefile::Parser's POD.

0.210 2008-03-16
* Added support for multiple single-colon rules to Makefile::Parser.
* Added support for multiple double-colon rules to Makefile::parser.

0.209 2008-03-12
* Calling external GNU make via "make" rather than "/usr/bin/env make".

0.208 2008-03-10
* Added $VERSION to Makefile::AST and Makefile::AST::Evaluator.
* Always calling '/usr/bin/env make' in makesimple.

0.207 2008-03-10
* Added POD for Makefile::AST.
* Added POD for Makefile::AST::Evaluator.
* Added POD for pgmake-db.

0.206 2008-03-10
* Added POD for Makefile::Parser::GmakeDB.
* Added POD for the makesimple script.
* Added pointers for the GmakeDB parser to Makefile::Parser's POD.

0.205 2008-03-10
* Removed bogus debugging code from the makesimple script.
* Fixed the test suite a bit.

0.204 2008-03-08
* Added missing external dependencies to Makefile.PL

0.203 2008-03-08
* Added File::Slurp to Makefile.PL

0.202 2008-03-08
* Added missing dependencies to Makefile.PL
* updated inc/ using the latest Module::Install

0.201 2008-03-08
* Added t/lib from the mdom repos.

0.20 2008-03-07
* added Makefile::Parser::GmakeDB and Makefile::AST

0.17 2007-03-16
* added support for the builtins 'error', 'warning', and 'info'

0.16 2007-03-16
* the parser now removes comments in variable assignments
* added support for the following GNU make builtin functions:
** if
** and
** or
** shell
** foreach

* removed /r from Parsre.pm
* added support for the shell function
* degraded the version of Test::More that Makefile::Parser requires

0.15 2007-03-14
* fixed the "too-many-args" issue
* added support for the following GNU make makefile builtin functions:
** word
** wordlist
** words
** firstword
** lastword
** dir
** notdir
** suffix
** basename
** addprefix
** join
** wildcard
** realpath
** abspath
* added _split_args method to do proper function arugment splitting
* refactored the second half of _process_refs out to _solve_refs_in_tokens
* merged the code of _check_func_args into _split_args
* _split_args now splits func arguments lazily
* updated the POD to reflect recent changes

0.14 2007-03-10
* added some more POD to plmake

0.13 2007-03-10
* minor POD fixes: C<< ... >> and SVN repos URL

0.12 2007-03-10
* added the C<prereqs> method to Makefile::Parser
* added POD accordingly (deprecated C<depends>)
* explaned the current status and plans for this module in the POD
* checked in a naive "make" command-line utility "plmake" which uses Makefile::Parser
* added basic support for commands after ';' in rules
* added a second optional argument to the C<parse> method so as to pass initial variable settings to the parser
* updated plmake to pass %ENV and variables specified on the command-line to the parser
* forced to use "/bin/sh" in run_commands
* implemented GNU make function 'subst', 'patsubst', 'strip', 'findstring', 'filter', 'filter-out' and 'sort'.
* implemented the Substitution References in GNU makefiles, for example, $(objects:.o=.c)
* fixed a bug regarding single-letter variable expansion

0.11 10/17/2005
* User-defined variable names are defined as /\w+/
* GNUMake's variable expansion sytax
        ${FOO}
  is now supported. Guretz++
* Add support for variable definition sytax
        FOO := blah blah blah
* Expand $@ in commands as expected
* Optimize the parser effectively by following Guretz Maxim's
  enlightening suggestion. Guretz++

0.10 10/16/2005
* Update the version number to 0.10
* Fix the platform-specific tests, stripped CR from test files,
  and test both against Cygwin and Win32. Slaven++

0.09 10/15/2005
* Fix a stupid bug in Parser.pm. The order of the two suffixes is
  inverted accidentally.
* Update the POD document, implementation, and tests accordingly.

0.08 10/10/2005
* Make Makefile::Parser->parse to raise syntax errors under strict mode
* Add full support for implicit pattern rules:
    %.o : %.c
        $(CC) -c $<
  The order of the implicit rules is not significant. Whee!
* Add more tests for pattern rules, expand $< and $* as expected, and clean up
  the stderr output.
* Add support for implicit suffix rules:
    .c.o:
        $(CC) -c $<
  Currently .SUFFIXES is a no-op. So suffix rules will be applied to any 
  suffix-like targets.
  Internally the parser converts the suffix rules to pattern
  rules, thus saving me a lot of coding.

0.07 10/5/2005
* Host this module to a SVN repository at OSSF

lib/Makefile/Parser/GmakeDB.pm
package Makefile::Parser::GmakeDB;

use strict;
use warnings;

#use Smart::Comments '####';
#use Smart::Comments '###', '####';
use List::Util qw( first );
use List::MoreUtils qw( none );
use MDOM::Document::Gmake;
use Makefile::AST;

our $VERSION = '0.215';

# XXX This should not be hard-coded this way...
our @Suffixes = (
    '.out',
    '.a',
    '.ln',
    '.o',
    '.c',
    '.cc',
    '.C',
    '.cpp',
    '.p',
    '.f',
    '.F',
    '.r',
    '.y',
    '.l',
    '.s',
    '.S',
    '.mod',
    '.sym',
    '.def',
    '.h',
    '.info',
    '.dvi',
    '.tex',
    '.texinfo',
    '.texi',
    '.txinfo',
    '.w',
    '.ch',
    '.web',
    '.sh',
    '.elc',
    '.el'
);

# need a better place for this sub:
sub solve_escaped ($) {
    my $ref = shift;
    $$ref =~ s/\\ ([\#\\:\n])/$1/gx;
}

sub _match_suffix ($@);

sub _match_suffix ($@) {
    my ($target, $full_match) = @_;
    ## $target
    ## $full_match
    if ($full_match) {
        return first { $_ eq $target } @Suffixes;
    } else {
        my ($fst, $snd);
        for my $suffix (@Suffixes) {
            my $len = length($suffix);
            ## prefix 1: substr($target, 0, $len)
            ## prefix 2: $suffix
            if (substr($target, 0, $len) eq $suffix) {
                $fst = $suffix;
                ## first suffix recognized: $suffix
                ## suffix 1: substr($target, $len)
                $snd = _match_suffix(substr($target, $len), 1);
                ## $snd
                next if !defined $snd;
                return ($fst, $snd);
            }
        }
        return undef;
    }
}

sub parse ($$) {
    shift;
    my $ast = Makefile::AST->new;
    my $dom = MDOM::Document::Gmake->new(shift);
    my ($var_origin, $orig_lineno, $orig_file);
    my $rule; # The last rule in the context
    my ($not_a_target, $directive);
    my $db_section = 'null';
    my $next_var_lineno = 0; # lineno for the next var assignment
    for my $elem ($dom->elements) {
        ## elem class: $elem->class
        ## elem lineno: $elem->lineno
        ## NEXT VAR LINENO: $next_var_lineno
        ## CURRENT LINENO: $elem->lineno
        if ($elem =~ /^# Variables$/) {
            ### Setting DB section to 'var': $elem->content
            $db_section = 'var';
            next;
        }
        if ($elem =~ /^# (?:Implicit Rules|Directives|Files)$/) {
            ### Setting DB section to 'rule': $elem->content
            $db_section = 'rule';
            next;
        }
        if ($elem =~ /^# (?:Pattern-specific Variable Values)$/) {
            ### Setting DB section to 'patspec': $elem->content
            $db_section = 'patspec';
            next;
        }
        if ($directive and $elem->class !~ /Directive$/) {
            # XXX yes, this is hacky
            ### pushing value to value: $elem
            push @{ $directive->{value} }, $elem->clone;
            next;
        }
        next if $elem->isa('MDOM::Token::Whitespace');
        if ($db_section eq 'var' and $elem->isa('MDOM::Assignment')) {
            ## Found assignment: $elem->source
            if (!$var_origin) {
                my $lineno = $elem->lineno;
                die "ERROR: line $lineno: No flavor found for the assignment";
            } else {
                my $lhs = $elem->lhs;
                my $rhs = $elem->rhs;
                my $op  = $elem->op;

                my $flavor;
                if ($op eq '=') {
                    $flavor = 'recursive';
                } elsif ($op eq ':=') {
                    $flavor = 'simple';
                } else {
                    # XXX add support for ?= and +=
                    die "Unknown op: $op";
                }
                my $name = join '', @$lhs; # XXX solve refs?
                my @value_tokens = map { $_->clone } @$rhs;
                #map { $_ = "$_" } @$rhs;
                ## LHS: $name
                ## RHS: $rhs
                my $var = Makefile::AST::Variable->new({
                    name   => $name,
                    flavor => $flavor,
                    origin => $var_origin,
                    value  => \@value_tokens,
                    lineno => $orig_lineno,
                    file => $orig_file,
                });
                $ast->add_var($var);
                undef $var_origin;
            }
        }
        elsif ($elem =~ /^#\s+(automatic|makefile|default|environment|command line)/) {
            $var_origin = $1;
            $var_origin = 'file' if $var_origin eq 'makefile';
            $next_var_lineno = $elem->lineno + 1;
        }
        elsif ($elem =~ /^# `(\S+)' directive \(from `(\S+)', line (\d+)\)/) {
            ($var_origin, $orig_file, $orig_lineno) = ($1, $2, $3);
            $next_var_lineno = $elem->lineno + 1;
            ### directive origin: $var_origin
            ### directive lineno: $orig_lineno
        }
        elsif ($elem =~ /^#\s+.*\(from `(\S+)', line (\d+)\)/) {
            ($orig_file, $orig_lineno) = ($1, $2);
            ## lineno: $orig_lineno
        }
        elsif ($db_section eq 'rule' and $elem =~ /^# Not a target:$/) {
            $not_a_target = 1;
        }
        elsif ($elem =~ /^#  Implicit\/static pattern stem: `(\S+)'/) {
            #### Setting pattern stem for solved implicit rule: $1
            $rule->{stem} = $1;
        }
        elsif ($db_section eq 'rule' and $elem =~ /^#  Also makes: (.*)/) {
            my @other_targets = split /\s+/, $1;
            $rule->{other_targets} = \@other_targets;
            #### Setting other targets: @other_targets
        }
        elsif ($db_section ne 'var' and
            $next_var_lineno == $elem->lineno and
            $elem =~ /^# (\S.*?) (:=|\?=|\+=|=) (.*)/) {
            #die "HERE!";
            my ($name, $op, $value) = ($1, $2, $3);
            # XXX tokenize the $value here?
            if (!$rule) {
                die "error: target/parttern-specific variables found where there is no rule in the context";
            }
            my $flavor;
            if ($op eq ':=') {
                $flavor = 'simple';
            } else {
                # XXX we should treat '?=' and '+=' specifically here?
                $flavor = 'recursive';
            }
            #### Adding local variable: $name
            my $handle = sub {
                my $ast = shift;
                my $old_value = $ast->eval_var_value($name);
                #warn "VALUE!!! $value";
                $value = "$old_value $value" if $op eq '+=';
                my $var = Makefile::AST::Variable->new({
                    name => $name,
                    flavor => $flavor,
                    origin => $var_origin,
                    value => $flavor eq 'recursive' ? $value : [$value],
                    lineno => $orig_lineno,
                    file => $orig_file,
                });
                $ast->enter_pad();
                $ast->add_var($var);
            };
            $ast->add_pad_trigger($rule->target => $handle);
            undef $var_origin;
        }
        elsif ($db_section ne 'var' and $elem->isa('MDOM::Rule::Simple')) {
            ### Found rule: $elem->source
            ### not a target? : $not_a_target
            if ($rule) {
                # The db output tends to produce
                # trailing empty commands, so we remove it:
                if ($rule->{commands}->[-1] and
                      $rule->{commands}->[-1] eq "\n") {
                    pop @{ $rule->{commands} };
                }
            }
            if ($not_a_target) {
                $not_a_target = 0;
                next;
            }
            my $targets = $elem->targets;
            my $colon   = $elem->colon;
            my $normal_prereqs = $elem->normal_prereqs;
            my $order_prereqs = $elem->order_prereqs;
            my $command = $elem->command;

            ## Target (raw): $targets
            ## Prereq (raw): $prereqs

            my $target = join '', @$targets;
            my @order_prereqs =  split /\s+/, join '', @$order_prereqs;
            my @normal_prereqs =  split /\s+/, join '', @$normal_prereqs;

            # Solve escaped chars:
            solve_escaped(\$target);
            map { solve_escaped(\$_) } @normal_prereqs, @order_prereqs;
            @order_prereqs = grep {
                my $value = $_;
                none { $_ eq $value } @normal_prereqs
            } @order_prereqs if @normal_prereqs;

            #### Target: $target
            ### Normal Prereqs: @normal_prereqs
            ### Order-only Prereqs: @order_prereqs

            #map { $_ = "$_" } @normal_prereqs, @order_prereqs;
            # XXX suffix rules allow order-only prepreqs? not sure...
            if ($target !~ /\s/ and $target !~ /\%/ and !@normal_prereqs and !@order_prereqs) {
                ## try to recognize suffix rule: $target
                my ($fst, $snd);
                $fst = _match_suffix($target, 1);
                if (!defined $fst) {
                    ($fst, $snd) = _match_suffix($target);
                    ## got first: $fst
                    ## got second: $snd
                    if (defined $fst) {
                        ## found suffix rule/2: $target
                        $target = '%' . $snd;
                        @normal_prereqs = ('%' . $fst);
                    }
                } else {
                    ## found suffix rule rule/1: $target
                    $target = '%' . $fst;
                }
            }
            my $rule_struct = {
                order_prereqs => [],
                normal_prereqs => \@normal_prereqs,
                order_prereqs => \@order_prereqs,
                commands => [defined $command ? $command : ()],
                colon => $colon,
                target => $target,
            };
            if ($target =~ /\%/) {
                ## implicit rule found: $target
                my $targets = [split /\s+/, $target];
                $rule_struct->{targets} = $targets,
                $rule = Makefile::AST::Rule::Implicit->new($rule_struct);
                $ast->add_implicit_rule($rule) if $db_section eq 'rule';
            } else {
                $rule = Makefile::AST::Rule->new($rule_struct);
                $ast->add_explicit_rule($rule) if $db_section eq 'rule';
            }
        } elsif ($elem->isa('MDOM::Command')) {
            ## Found command: $elem
            if (!$rule) {
                die "error: line " . $elem->lineno .
                    ": Command not allowed here";
            } else {
                #my @tokens = map { "$_" } $elem->elements;
                #my @tokens = $elem
                #shift @tokens if $tokens[0] eq "\t";
                #pop @tokens if $tokens[-1] eq "\n";
                #push @{ $rule->{commands} }, \@tokens;
                ## parser: CMD: $elem
                my $first = $elem->first_element;
                ## $first
                $elem->remove_child($first)
                    if $first->class eq 'MDOM::Token::Separator';
                ### elem source: $elem->source
                #if ($elem->source eq "\n") {
                #    die "Matched!";
                #}
                ## lineno2: $orig_lineno
                $elem->{lineno} = $orig_lineno if $orig_lineno;
                $rule->add_command($elem->clone); # XXX why clone?
                ### Command added: $elem->content
            }
        } elsif ($elem->class =~ /MDOM::Directive/) {
            ### directive name: $elem->name
            ### directive value: $elem->value
            if ($elem->name eq 'define') {
                # XXX set lineno to $orig_lineno here?
                $directive = {
                    name => $elem->value,
                    value => [], # needs to be fed later
                    flavor => 'recursive',
                    origin => $var_origin,
                    lineno => $orig_lineno,
                    file => $orig_file,
                };
                next;
            }
            if ($elem->name eq 'endef') {
                ### parsed a define directive: $directive
                # trim the trailing new lines in the value:
                #warn "HERE!!! ";
                #warn quotemeta($directive->{value}->[-1]);
                my $last = $directive->{value}->[-1];
                if ("$last" =~ /^\s*$/s) {
                    pop @{ $directive->{value} }

                } elsif ($last->can('last_element')) {
                    my $last_elem = $last->last_element;
                    #warn "LAST: '$last'\n";
                    if ($last_elem and "$last_elem" =~ /^\s*$/s) {
                        $last->remove_child($last_elem);
                    }
                }

                my $var = Makefile::AST::Variable->new($directive);
                $ast->add_var($var);
                undef $var_origin;
                undef $directive;

            } else {
                warn "warning: line " . $elem->lineno .
                    ": Unknown directive: " . $elem->source;
            }
        } elsif ($elem->class =~ /Unknown/) {
            # XXX Note that output from $(info ...) may skew up stdout
            # XXX This hack is used to make features/conditionals.t pass
            print $elem if $elem eq "success\n";
            # XXX The 'hello, world' hack to used to make sanity/func-refs.t pass
            warn "warning: line " . $elem->lineno .
                ": Unknown GNU make database struct: " .
                $elem->source
                if $elem !~ /hello.*world/ and
                   $elem ne "success\n";
        }
    }
    {
        my $default = $ast->eval_var_value('.DEFAULT_GOAL');
        ## default goal's value: $var
        $ast->default_goal($default) if $default;
        ### DEFAULT GOAL: $ast->default_goal

        my $rule = $ast->apply_explicit_rules('.PHONY');
        if ($rule) {
            ### PHONY RULE: $rule
            ### phony targets: @{ $rule->normal_prereqs }
            for my $phony (@{ $rule->normal_prereqs }) {
                $ast->set_phony_target($phony);
            }
        }
        ## foo var: $ast->get_var('foo')
    }
    $ast;
}

1;
__END__

=head1 NAME

Makefile::Parser::GmakeDB - GNU makefile parser using GNU make's database dump

=head1 VERSION

This document describes Makefile::Parser::GmakeDB 0.215 released on 18 August 2011.

=head1 SYNOPSIS

    use Makefile::Parser::GmakeDB;
    my $db_listing = `make --print-data-base -pqRrs -f Makefile`;
    my $ast = Makefile::Parser::GmakeDB->parse(\$db_listing);

=head1 DESCRIPTION

This module serves as a parser for GNU makefiles. However, it does not parse
user's original makefile directly. Instead it uses L<Makefile::DOM> to parse
the "data base output listing" produced by GNU make (via its
C<--print-data-base> option). So essentially it reuses the C
implementation of GNU make.

This parser has been tested as a component of the L<pgmake-db> utility and has successfully passed 51% of GNU make 3.81's official test suite.

The result of the parser is a makefile AST defined by L<Makefile::AST>.

The "data base output listing" generated by C<make --print-data-base> is a detailed listing for GNU make's internal data structures, which is essentially the AST used by C<make>. According to GNU make's current maintainer, Paul Smith, this feature is provided primarily for debuging the user's own makefiles, and it also helps the GNU make developer team to diagnose the flaws in make itself. Incidentally this output is conformed to the GNU makefile syntax, and a lot of important information is provided in the form of makefile comments. Therefore, my GmakeDB parser is able to reuse the L<Makefile::DOM> module to parse this output listing.

The data base output from GNU make can be divided into several clearly-separated segments. They're file header, "Variables", "Files", "VPATH Search Paths", as well as the last resource stats information.

The contents of these segments are mostly obvious. The Files segment may deserve some explanation. It is the place for explict rules.

Now let's take the Variables segment as an  example to demonstrate the format of the data base listing:

    # Variables

    # automatic
    <D = $(patsubst %/,%,$(dir $<))
    # automatic
    ?F = $(notdir $?)
    # environment
    DESKTOP_SESSION = default
    # automatic
    ?D = $(patsubst %/,%,$(dir $?))
    # environment
    GTK_RC_FILES = /etc/gtk/gtkrc:/home/agentz/.gtkrc-1.2-gnome2
    # environment
    ...

It's shown that the flavor and origin of the makefile variables are given in the previous line as comments. Hence feeding this back into GNU make again makes little sense.

Similarly, the Files segment for explicit rules also puts big amount of the important information into makefile comments:

    # Files

    # Not a target:
    bar.c:
    #  Implicit rule search has not been done.
    #  Modification time never checked.
    #  File has not been updated.

    all: foo.o bar.o
    #  Implicit rule search has been done.
    #  File does not exist.
    #  File has not been updated.
    # variable set hash-table stats:
    # Load=0/32=0%, Rehash=0, Collisions=0/0=0%

    foo.o: foo.c
    #  Implicit rule search has not been done.
    #  Implicit/static pattern stem: `foo'
    #  File does not exist.
    #  File has not been updated.
    # variable set hash-table stats:
    # Load=0/32=0%, Rehash=0, Collisions=0/0=0%
    #  commands to execute (from `ex2.mk', line 8):
        $(CC) -c $(CFLAGS) $< -o $@
    ...

From the previous two data base listing snippets, it's not hard to see that the variable references in rule commands and recursively-expanded variables's values are not expanded.

Experiments have shown that GNU make will do implicit rule search for the first rule that needs to, but no more. This behavior means testing our own implicit rule searching algorithm requires specifying at least two goals that require matching.

=head1 DEPENDENCIES

=over

=item GNU make 3.81

At least the F<make> executable of GNU make 3.81 is required to work with this module.

=item L<Makefile::DOM>

=back

=head1 BUGS

=over

=item *

GNU make does not escape meta characters appeared in rule targes and prerequisites in its data base listing. Examples are C<:>, C<\>, and C<#>. This bug has been reported to the GNU make team as C<Savannah bug #20067>.

This bug has not yet been fixed on the C<make> side, so I have to work around this issue by preprocessing the data base listing in the L<makesimple> script.

=item *

The data base listing produced by GNU make lacks the information regarding the C<export> and C<unexport> directives. It gives rise to the lack of information in the resulting AST structures constructed by this module. Hence the current AST and runtime do not implement the C<export> and C<unexport> directives.

To make it even worse, there's no known way to work around it.

I've already reported this issue to the GNU make team as Savannah bug #20069.

=back

=head1 CODE REPOSITORY

For the very latest version of this script, check out the source from

L<http://github.com/agentzh/makefile-parser-pm>.

There is anonymous access to all.

=head1 AUTHOR

Zhang "agentzh" Yichun C<< <agentzh@gmail.com> >>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2005-2008 by Zhang "agentzh" Yichun (agentzh).

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<Makefile::AST>, L<Makefile::AST::Evaluator>, L<Makefile::DOM>,
L<makesimple>, L<pgmake-db>.

